File: packages/launcher/shared/mc_metadata/src/version_json.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VersionJson {
    pub id: String,
    #[serde(rename = "type")]
    pub version_type: String,
    #[serde(rename = "mainClass")]
    pub main_class: String,
    #[serde(rename = "minecraftArguments")]
    pub minecraft_arguments: Option<String>,
    pub arguments: Option<Arguments>,
    pub libraries: Vec<Library>,
    #[serde(rename = "assetIndex")]
    pub asset_index: Option<AssetIndex>,
    pub assets: Option<String>,
    pub downloads: Option<VersionDownloads>,
    #[serde(rename = "inheritsFrom")]
    pub inherits_from: Option<String>,
    #[serde(rename = "jar")]
    pub jar: Option<String>,
    #[serde(rename = "releaseTime")]
    pub release_time: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VersionDownloads {
    pub client: Option<ClientDownload>,
    pub server: Option<ServerDownload>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientDownload {
    pub sha1: Option<String>,
    pub size: Option<u64>,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerDownload {
    pub sha1: Option<String>,
    pub size: Option<u64>,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Arguments {
    pub game: Option<Vec<ArgumentValue>>,
    pub jvm: Option<Vec<ArgumentValue>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ArgumentValue {
    String(String),
    Conditional {
        rules: Vec<Rule>,
        value: ArgumentData,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ArgumentData {
    Single(String),
    Multiple(Vec<String>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rule {
    pub action: String,
    pub os: Option<OsRule>,
    pub features: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OsRule {
    pub name: Option<String>,
    pub version: Option<String>,
    pub arch: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetIndex {
    pub id: String,
    pub sha1: Option<String>,
    pub size: Option<u64>,
    #[serde(rename = "totalSize")]
    pub total_size: Option<u64>,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Library {
    pub name: String,
    pub downloads: Option<LibraryDownloads>,
    pub natives: Option<serde_json::Value>,
    pub extract: Option<ExtractRule>,
    pub rules: Option<Vec<Rule>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LibraryDownloads {
    pub artifact: Option<Artifact>,
    pub classifiers: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Artifact {
    pub path: String,
    pub sha1: Option<String>,
    pub size: Option<u64>,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractRule {
    pub exclude: Option<Vec<String>>,
}

impl VersionJson {
    pub fn load(path: &std::path::Path) -> anyhow::Result<Self> {
        let data = std::fs::read_to_string(path)?;
        let vj: VersionJson = serde_json::from_str(&data)?;
        Ok(vj)
    }

    pub fn load_version(version_id: &str) -> anyhow::Result<Self> {
        let path = _shared_core::platform::versions_dir()
            .join(version_id)
            .join(format!("{version_id}.json"));
        Self::load(&path)
    }

    pub fn collect_all_args(&self) -> (Vec<String>, Vec<String>) {
        let mut game_args = Vec::new();
        let mut jvm_args = Vec::new();

        if let Some(ref args) = self.arguments {
            if let Some(ref game) = args.game {
                for av in game {
                    match av {
                        ArgumentValue::String(s) => game_args.push(s.clone()),
                        ArgumentValue::Conditional { rules, value } => {
                            if should_apply(rules) {
                                match value {
                                    ArgumentData::Single(s) => game_args.push(s.clone()),
                                    ArgumentData::Multiple(v) => {
                                        game_args.extend(v.iter().cloned())
                                    }
                                }
                            }
                        }
                    }
                }
            }
            if let Some(ref jvm) = args.jvm {
                for av in jvm {
                    match av {
                        ArgumentValue::String(s) => jvm_args.push(s.clone()),
                        ArgumentValue::Conditional { rules, value } => {
                            if should_apply(rules) {
                                match value {
                                    ArgumentData::Single(s) => jvm_args.push(s.clone()),
                                    ArgumentData::Multiple(v) => jvm_args.extend(v.iter().cloned()),
                                }
                            }
                        }
                    }
                }
            }
        }

        if let Some(ref mc_args) = self.minecraft_arguments {
            game_args.extend(mc_args.split_whitespace().map(String::from));
        }

        (jvm_args, game_args)
    }
}

fn should_apply(rules: &[Rule]) -> bool {
    let mut allow = false;
    let mut deny = false;

    for rule in rules {
        let os_match = match &rule.os {
            Some(os) => {
                let name_ok = match &os.name {
                    Some(name) => match name.as_str() {
                        "windows" => cfg!(windows),
                        "linux" => cfg!(target_os = "linux"),
                        "osx" => cfg!(target_os = "macos"),
                        _ => true,
                    },
                    None => true,
                };
                let arch_ok = match &os.arch {
                    Some(arch) => match arch.as_str() {
                        "x86" => cfg!(target_arch = "x86"),
                        _ => true,
                    },
                    None => true,
                };
                name_ok && arch_ok
            }
            None => true,
        };

        if os_match {
            match rule.action.as_str() {
                "allow" => allow = true,
                "disallow" => deny = true,
                _ => {}
            }
        }
    }

    if rules.is_empty() {
        return true;
    }
    allow && !deny
}

File: packages/minecraft-mod-mcp/src/mc/version-json.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { isWindows, isMacos } from "../runtime/detector.js";
import { versionsDir, librariesDir } from "./platform.js";

export interface VersionJson {
  id: string;
  type: string;
  mainClass: string;
  minecraftArguments?: string;
  arguments?: {
    game?: ArgumentValue[];
    jvm?: ArgumentValue[];
  };
  libraries: Library[];
  assetIndex?: AssetIndex;
  assets?: string;
  downloads?: VersionDownloads;
  inheritsFrom?: string;
  jar?: string;
  releaseTime?: string;
}

export type ArgumentValue = string | ConditionalArgument;

export interface ConditionalArgument {
  rules: Rule[];
  value: string | string[];
}

export interface Rule {
  action: string;
  os?: OsRule;
  features?: Record<string, unknown>;
}

export interface OsRule {
  name?: string;
  version?: string;
  arch?: string;
}

export interface AssetIndex {
  id: string;
  sha1?: string;
  size?: number;
  totalSize?: number;
  url?: string;
}

export interface VersionDownloads {
  client?: ClientDownload;
  server?: ServerDownload;
}

export interface ClientDownload {
  sha1?: string;
  size?: number;
  url?: string;
}

export interface ServerDownload {
  sha1?: string;
  size?: number;
  url?: string;
}

export interface Library {
  name: string;
  downloads?: LibraryDownloads;
  natives?: Record<string, string>;
  extract?: { exclude?: string[] };
  rules?: Rule[];
}

export interface LibraryDownloads {
  artifact?: Artifact;
  classifiers?: Record<string, Record<string, unknown>>;
}

export interface Artifact {
  path: string;
  sha1?: string;
  size?: number;
  url?: string;
}

export function loadVersionJson(path: string): VersionJson {
  const data = readFileSync(path, "utf-8");
  return JSON.parse(data) as VersionJson;
}

export function loadVersion(versionId: string): VersionJson {
  const dir = versionsDir();
  const path = join(dir, versionId, `${versionId}.json`);
  if (!existsSync(path)) {
    throw new Error(
      `Version "${versionId}" is not installed.\n` +
      `  Missing: ${path}\n` +
      `  Run: minecraft-mod-mcp install ${versionId}`
    );
  }
  return loadVersionJson(path);
}

export function loadVersionMerged(versionId: string): VersionJson {
  const primary = loadVersion(versionId);
  return mergeWithParents(primary);
}

function mergeWithParents(vj: VersionJson): VersionJson {
  if (!vj.inheritsFrom) return vj;

  let parent: VersionJson;
  try {
    parent = loadVersion(vj.inheritsFrom);
  } catch {
    return vj;
  }

  parent = mergeWithParents(parent);

  return {
    id: vj.id,
    type: vj.type || parent.type,
    mainClass: vj.mainClass || parent.mainClass,
    inheritsFrom: vj.inheritsFrom,
    libraries: mergeLibraries(parent.libraries ?? [], vj.libraries ?? []),
    arguments: mergeArguments(parent.arguments, vj.arguments),
    minecraftArguments: vj.minecraftArguments || parent.minecraftArguments,
    assetIndex: vj.assetIndex || parent.assetIndex,
    assets: vj.assets || parent.assets,
    downloads: vj.downloads || parent.downloads,
    jar: vj.jar || parent.jar,
    releaseTime: vj.releaseTime || parent.releaseTime,
  };
}

function libBaseKey(name: string): string {
  const parts = name.split(":");
  return parts.length >= 2 ? `${parts[0]}:${parts[1]}` : name;
}

function mergeLibraries(parent: Library[], child: Library[]): Library[] {
  const map = new Map<string, Library>();
  for (const lib of parent) {
    if (lib.name) map.set(libBaseKey(lib.name), lib);
  }
  for (const lib of child) {
    if (lib.name) map.set(libBaseKey(lib.name), lib);
  }
  return [...map.values()];
}

function mergeArguments(
  parent: VersionJson["arguments"],
  child: VersionJson["arguments"],
): VersionJson["arguments"] {
  if (!parent && !child) return undefined;
  if (!parent) return child;
  if (!child) return parent;
  return {
    game: [...(parent.game ?? []), ...(child.game ?? [])],
    jvm: [...(parent.jvm ?? []), ...(child.jvm ?? [])],
  };
}

export function collectAllArgs(vj: VersionJson): { jvmArgs: string[]; gameArgs: string[] } {
  const gameArgs: string[] = [];
  const jvmArgs: string[] = [];

  if (vj.arguments) {
    if (vj.arguments.game) {
      for (const av of vj.arguments.game) {
        if (typeof av === "string") {
          gameArgs.push(av);
        } else if (shouldApply(av.rules)) {
          if (typeof av.value === "string") gameArgs.push(av.value);
          else gameArgs.push(...av.value);
        }
      }
    }
    if (vj.arguments.jvm) {
      for (const av of vj.arguments.jvm) {
        if (typeof av === "string") {
          jvmArgs.push(av);
        } else if (shouldApply(av.rules)) {
          if (typeof av.value === "string") jvmArgs.push(av.value);
          else jvmArgs.push(...av.value);
        }
      }
    }
  }

  if (vj.minecraftArguments) {
    gameArgs.push(...vj.minecraftArguments.split(/\s+/));
  }

  return { jvmArgs, gameArgs };
}

export function shouldApply(rules: Rule[], features?: Record<string, boolean>): boolean {
  if (rules.length === 0) return true;
  const feats = features ?? {};

  let allow = false;
  let deny = false;

  for (const rule of rules) {
    if (rule.features) {
      const featMatch = Object.entries(rule.features).every(([k, v]) => {
        if (typeof v === "boolean") return feats[k] === v;
        return feats[k] != null;
      });
      if (!featMatch) continue;
    }

    const osMatch = rule.os ? matchOs(rule.os) : true;
    if (osMatch) {
      if (rule.action === "allow") allow = true;
      else if (rule.action === "disallow") deny = true;
    }
  }

  return allow && !deny;
}

function matchOs(os: OsRule): boolean {
  if (os.name) {
    const nameOk = (() => {
      switch (os.name) {
        case "windows": return isWindows();
        case "linux": return !isWindows() && !isMacos();
        case "osx": return isMacos();
        default: return true;
      }
    })();
    if (!nameOk) return false;
  }
  return true;
}

export function libraryPath(name: string): string {
  const parts = name.split(":");
  if (parts.length < 3) return join(librariesDir(), name.replace(/:/g, "/"));

  const group = parts[0].replace(/\./g, "/");
  const artifact = parts[1];
  const version = parts[2];
  const classifier = parts[3];

  const filename = classifier
    ? `${artifact}-${version}-${classifier}.jar`
    : `${artifact}-${version}.jar`;

  return join(librariesDir(), group, artifact, version, filename);
}

export function libraryMavenPath(name: string): string {
  const parts = name.split(":");
  if (parts.length < 3) return name.replace(/:/g, "/");
  const group = parts[0].replace(/\./g, "/");
  const artifact = parts[1];
  const version = parts[2];
  const classifier = parts[3];
  return classifier
    ? `${group}/${artifact}/${version}/${artifact}-${version}-${classifier}.jar`
    : `${group}/${artifact}/${version}/${artifact}-${version}.jar`;
}

export function resolveClasspath(libraries: Library[]): string[] {
  const classpath: string[] = [];

  for (const lib of libraries) {
    if (lib.rules && !shouldApply(lib.rules)) continue;

    if (lib.downloads?.artifact) {
      const path = join(librariesDir(), lib.downloads.artifact.path);
      if (existsSync(path)) classpath.push(path);
      continue;
    }

    const path = libraryPath(lib.name);
    if (existsSync(path)) {
      classpath.push(path);
    } else {
      const universalPath = path.replace(/\.jar$/, "-universal.jar");
      if (existsSync(universalPath)) classpath.push(universalPath);
    }
  }

  const patched = patchClasspath(classpath);

  return patched;
}

function patchClasspath(classpath: string[]): string[] {
  return classpath.map(p => {
    if (p.includes("launchwrapper-1.9")) {
      const better = p
        .replace(/launchwrapper[\\/]1\.9[\\/]/, "launchwrapper/1.12/")
        .replace("launchwrapper-1.9.jar", "launchwrapper-1.12.jar");
      if (existsSync(better)) return better;
    }
    return p;
  });
}
